src/Rule/ForbidNotNormalizedTypeRule.php

```diff
@@ -0,0 +1,564 @@
+                    $multiTypeNode,
+                    $nameSpace,
+                    "parameter {$paramTagValue->parameterName}",
+                    $this->getPhpDocLine($sourceNode, $paramTagValue),
+                );
+                $errors = array_merge($errors, $newErrors);
+            }
+        }
+
+        return $errors;
+    }
+
+    /**
+     * @param array<VarTagValueNode> $varTagValues
+     * @return list<RuleError>
+     */
+    public function processVarTags(
+        PhpParserNode $originalNode,
+        array $varTagValues,
+        NameScope $nameSpace
+    ): array
+    {
+        $errors = [];
+
+        foreach ($varTagValues as $varTagValue) {
+            foreach ($this->extractUnionAndIntersectionPhpDocTypeNodes($varTagValue->type) as $multiTypeNode) {
+                $identification = $varTagValue->variableName !== ''
+                    ? "variable {$varTagValue->variableName}"
+                    : null;
+
+                $newErrors = $this->processMultiTypePhpDocNode(
+                    $multiTypeNode,
+                    $nameSpace,
+                    $identification,
+                    $this->getPhpDocLine($originalNode, $varTagValue),
+                );
+                $errors = array_merge($errors, $newErrors);
+            }
+        }
+
+        return $errors;
+    }
+
+    /**
+     * @param array<ReturnTagValueNode> $returnTagValues
+     * @return list<RuleError>
+     */
+    public function processReturnTags(
+        PhpParserNode $originalNode,
+        array $returnTagValues,
+        NameScope $nameSpace
+    ): array
+    {
+        $errors = [];
+
+        foreach ($returnTagValues as $returnTagValue) {
+            foreach ($this->extractUnionAndIntersectionPhpDocTypeNodes($returnTagValue->type) as $multiTypeNode) {
+                $newErrors = $this->processMultiTypePhpDocNode($multiTypeNode, $nameSpace, 'return', $this->getPhpDocLine($originalNode, $returnTagValue));
+                $errors = array_merge($errors, $newErrors);
+            }
+        }
+
+        return $errors;
+    }
+
+    /**
+     * @return list<UnionTypeNode|IntersectionTypeNode>
+     */
+    private function extractUnionAndIntersectionPhpDocTypeNodes(TypeNode $typeNode): array
+    {
+        $nodes = [];
+        $this->traversePhpDocTypeNode($typeNode, static function (TypeNode $typeNode) use (&$nodes): void {
+            if ($typeNode instanceof UnionTypeNode || $typeNode instanceof IntersectionTypeNode) {
+                $nodes[] = $typeNode;
+            }
+
+            if ($typeNode instanceof NullableTypeNode) {
+                $nodes[] = new UnionTypeNode([$typeNode->type, new IdentifierTypeNode('null')]);
+            }
+        });
+        return $nodes;
+    }
+
+    /**
+     * @return list<IntersectionType|UnionType>
+     */
+    private function extractUnionIntersectionPhpParserNodes(PhpParserNode $node): array
+    {
+        $extractor = new UnionIntersectionExtractorVisitor();
```

Comment:
This is a bit overkill. Because PHP native types are always in DNF form, they can be extracted without deep recursion.

Reply:
Good idea, thx.